Perform SSA variable renaming.

vext01 opened the pull request:
Please check carefully. I've been working on this code for so long I feel I can barely see the wood for the trees!

There's a ykpack change associated with this PR. Coming soon.

We perform renaming in-place using a slightly modified version of the
algorithm from the book: Modern Compiler Implementation in Java, 2nd
edition, Andrew Appel.

We experimented with the algorithm from the SSA book by Jeremy Singer
but couldn't get it to work. I suspect there is an error in the
published algorithm, but I may be wrong.

For now this only renames variables for implemented TIR bytecodes, and
not yet in terminators (so return terminators currently always return
variable 0, which will need to be changed).

It may be possible to have the variables numbers start from zero in the
resulting bytecode, but I'm keen to get this in first.

Here are some example CFGs:

![image](https://user-images.githubusercontent.com/604955/56585107-1407b680-65d5-11e9-8c8a-96dce49dd471.png)

---

![image](https://user-images.githubusercontent.com/604955/56585134-21bd3c00-65d5-11e9-9740-13b07144c31e.png)

vext01 commented:
I wonder if I should add comments where I had previously changed the codegen traits?

vext01 commented:
Also note that I did not gate off the MIR serialiser, since I thin kwe will need this for the software tracer too.

ltratt commented:
> I wonder if I should add comments where I had previously changed the codegen traits?

Up to you.

ltratt commented:
bors r+

ltratt commented:
bors r-

ltratt commented:
bors r+

vext01 commented:
`gdb not available but debuginfo gdb debuginfo test requested`

I've installed it on b12, please ask bors to retry.

ltratt commented:
bors r+

vext01 commented:
I don't think this is my change causing this (but I will test on master).

This *might* be:
https://github.com/rust-lang/rust/issues/52452

But we use gdb-7.12, not 8.1.

vext01 commented:
> I don't think this is my change causing this (but I will test on master).

These failures do indeed happen on master. I will now try some different gdb versions.

ltratt commented:
OK.

vext01 commented:
Well I'm confused.

I've tried gdb versions:
 * 7.11.1
 * 7.12
 * 8.2

And they all have gdb compiletest failures on upstream master. Upstream are using 8.2 it seems.

I also tried updating master to track upstream and repeating the tests. No joy.

ltratt commented:
What do you want to do?

vext01 commented:
I'm not sure. Ideally the tests would pass!

I've asked what's going on in discord.

In the meantime, we could try to disable those tests (if possible)?

vext01 commented:
Let me try once more using a fresh clone of rust-lang/rust. Master on our fork has merge commits that are not present upstream. Not that I think that should make a blind bit of difference, but if it still shows, we can raise a bug upstream, or reopen this:

https://github.com/rust-lang/rust/issues/52452

ltratt commented:
OK.

vext01 commented:
Yep, I was able to reproduce this on upstream master. I've commented on the bug.

vext01 commented:
In the interest of moving on, we could do the following from the compiler guide, to "see if things are generally working correctly":

> ./x.py test --stage 1 src/test/{ui,compile-fail,run-pass}

That'd still be better than no tests.

vext01 commented:
Laurie, this should fix the failing tests.

We needed not only a newer, but a python-enabled, gdb.

That dependency is undocumented upstream, so I'll probably raise a PR against the compiler guide tomorrow.

If it looks OK to you, i'll squash and we can try the `r+` again.

ltratt commented:
bors try

vext01 commented:
After some experimentation last night, here's what's going on.

 * The tests work on upstream master
 * The tests do not work on our master
 * this PR branch is based on our master, and inherits test failures.

We need to sync our master and yk branch. Since that's a possibly non-trivial job, I'd like to do that as a separate PR.

For now we can enable a subset of the tests, if that's OK with you @ltratt.

ltratt commented:
Does that mean you need to push a change to this PR that disables some tests?

vext01 commented:
Correct.

ltratt commented:
OK, go ahead.

vext01 commented:
There's the commit. Give that a try. No need to squash just yet, as bors will no auto-merge a try command.

ltratt commented:
bors try

vext01 commented:
Squash?

ltratt commented:
Yep!

vext01 commented:
splat.

Bors will build need again after your `r+` due to the squash.

ltratt commented:
bors r+

vext01 committed:
Perform SSA variable renaming.

We perform renaming in-place using a slightly modified version of the
algorithm from the book: Modern Compiler Implementation in Java, 2nd
edition, Andrew Appel.

We experimented with the algorithm from the SSA book by Jeremy Singer
but couldn't get it to work. I suspect there is an error in the
published algorithm, but I may be wrong. I've email Jeremy.

For now this only renames variables for implemented TIR bytecodes, and
not yet in terminators (so return terminators currently always return
variable 0, which will need to be changed).

It may be possible to have the variables numbers start from zero in the
resulting bytecode, but I'm keen to get this in first.

vext01 commented on src/librustc_yk_sections/Cargo.toml:
```diff
@@ -12,7 +12,8 @@ test = false
 rustc_yk_link = { path = "../librustc_yk_link" }
-ykpack = { git = "https://github.com/softdevteam/ykpack" }
+#ykpack = { git = "https://github.com/softdevteam/ykpack" }
+ykpack = { path = "../../../ykpack" }
```
This will need to be revised once the ykpack change is in.

vext01 commented:
I just realised that ykpack is now in the mono repo. We could do with that transition being complete before I raise the companion PR to this PR.

ltratt commented:
> I just realised that ykpack is now in the mono repo. We could do with that transition being complete before I raise the companion PR to this PR.

The transition is complete isn't it?

ltratt commented on src/librustc_data_structures/graph/dominators/mod.rs:
```diff
@@ -118,6 +118,15 @@ impl<Node: Idx> Dominators<Node> {
     }
 
+    /// Find the children of a node in the dominator tree.
+    pub fn immediately_dominates(&self, node: Node) -> Vec<Node> {
```
I don't know if this will be possible (it probably depends on the borrow checker), but it would almost certainly be faster to make this an iterator rather than a producer of a `Vec`. You could probably get away with `-> impl Iterator<Node>` or similar.

vext01 replied:
There was a reason why I didn't do this, but I don't recall what it was. Let me experiment and come back to you.

vext01 commented:
> The transition is complete isn't it?

I don't think so. We have an outstanding PR here:
https://github.com/softdevteam/yk/pull/1

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -29,16 +34,26 @@ use std::io::Write;
 const SECTION_NAME: &'static str = ".yk_tir";
 const TMP_EXT: &'static str = ".yk_tir.tmp";
 
+/// The pre-SSA return value variable. In the pre-SSA TIR, we need the return value to reside at an
```
The first half of this comment makes sense, but the second half is confusing. Possibly it's just because some sentences are in the wrong order?

vext01 replied:
Can you expand on that -- it makes sense to me, but then I'm not a good critic for my own work.

ltratt replied:
I think you just need to try rephrasing it. I can't make sense of the comment as a whole at the moment.

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -57,25 +72,52 @@ struct ConvCx<'a, 'tcx, 'gcx> {
     /// Maps each block to the variable it defines. This is what Appel calls `A_{orig}`.
     block_defines: RefCell<IndexVec<BasicBlock, FxHashSet<TirLocal>>>,
     /// Monotonically increasing number used to give TIR variables a unique ID.
+    /// Note that a 0 is reserved for `PRE_SSA_RET_VAR`.
```
"a 0" -> "0"

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -57,25 +72,52 @@ struct ConvCx<'a, 'tcx, 'gcx> {
+    /// From the Appel book:
+    ///     "We consider the start node to contain an implicit definition of every variable,
+    ///     either because the variable may be a formal parameter or to represent the notion of
+    ///     a ← uninitialized without special cases.
```
I can't parse the part of the comment on this line.

vext01 replied:
Something has gone wrong somewhere. That's supposed to be \bottom. Will fix.

vext01 replied:
Actually I was wrong. This is verbatim from the book.

![image](https://user-images.githubusercontent.com/604955/56655161-e466b600-6689-11e9-9ebd-75b78bbe69e1.png)

vext01 replied:
maybe we can delimit the abstract algebra in some quotes for readability. I think the problem is `notion of a <- uninitialized` is hard to parse.

ltratt replied:
Agreed, maybe this just needs a bit of rephrasing / reformatting.

vext01 replied:
I added the quotes in the last commit. Does that work for you?

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -262,6 +306,108 @@ fn insert_phi(block: &mut ykpack::BasicBlock, var: TirLocal, arity: usize) {
+        let n_usize = n as usize;
+        // We have to remember the variables whose stacks we must pop from when we come back from
+        // recursion. These must be the variables *before* they were renamed.
+        let mut pop_later = Vec::new();
```
Can a variable appear more than once in here? Does it happen often that variables end up in here?

vext01 replied:
Every variable assignment in the block we are processing will push to this vector. The same variable may appear more than once if a block assigns to the same variable many times.

ltratt replied:
So this should probably be a set to avoid (I assume pointlessly) poping the same thing over and over again?

vext01 replied:
If a variable is assigned to X times, then we need to pop X times.

The the algorithm recursively walks the dominator tree depth-first. When it sees an assignment to a variable, it makes a new SSA variable, rewrites the old variable. and pushes the new variable onto the stack of the corresponding old variable. When we are bubbling up from recursion, we have to revert to the previous SSA versions of the variables. So if a block assigned to a variable `z` twice, we'd need to revert to `z` as of two assignments ago, thus popping twice.

Hard to explain...

ltratt replied:
Got it.

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -262,6 +306,108 @@ fn insert_phi(block: &mut ykpack::BasicBlock, var: TirLocal, arity: usize) {
+            // "Suppose n is the jth predecessor of y".
+            let j = mir.predecessors_for(y).iter().position(|b| b == &n_idx).unwrap();
+            for st in &mut blks[y.as_usize()].stmts {
+                match st.phi_arg_mut(j) {
```
`if let Some(ref mut a) = st.phi_arg_mut(j) { ... } `

vext01 commented:
This should address the comments.

Untested. Building now...

vext01 commented:
> Untested. Building now...

Our yk_tir test passes and manual inspection of the graphs show up nothing dodgy.

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -47,11 +47,11 @@ use rustc_data_structures::fx::FxHashSet;
+/// The pre-SSA return value variable. In MIR, a return terminator implicitly returns variable
+/// zero. We can't do this in TIR because TIR is in SSA form and the variable we return depends
+/// upon which SSA variable reaches the terminator. So, during initial TIR lowering, we convert the
+/// implicit MIR terminator to an explicit TIR terminator returning variable index zero initially.
```
I think we can remove the last  "initially".

ltratt commented on src/librustc_yk_sections/emit_tir.rs:
```diff
@@ -388,13 +388,12 @@ impl RenameCx {
-                    None => (), // It wasn't a Phi. Do nothing.
+                if let Some(ref mut a) = st.phi_arg_mut(j) {
+                    // We only get here if `st` was a Phi.
+                    let i = self.stack[**a as usize].last().cloned().unwrap();
```
`let i =` is longer than `**a =`. Can we use the latter?

vext01 replied:
We can, but I'd rather keep it as-is for parity with the algorithm in the book:

![image](https://user-images.githubusercontent.com/604955/56669086-84cbd300-66a8-11e9-824c-d7979a1b008d.png)

ltratt replied:
OK.

vext01 commented:
One commit to address your comment, another to give the instruction that does the SSA definitions a better name.

If you are happy, we can squash this ready for merge, but we cannot merge just yet. There's a yk change that I'm about to raise a PR for and that depends on the new tarball PR with cargo.

Sadly my stage 2 failed because I forgot to specify the config file on the command line and now I have to rebuild...

ltratt commented:
Please squash.

vext01 commented:
squashed. Now just waiting on the ykrustc archives. Once that's done, I'll flip the ykpack dependency over to yk.

vext01 committed:
Update Return generation in-line with ykpack change.

vext01 commented:
This is now irrelevant.
